Rejoin PCT in the Shell

zqsz-xiv opened the pull request:
_Let expanse contract, eon become instant..._

This PR merges PCT in the Shell back to BLM in the Shell. Deployments remain separate, and PCT is kept separate by manually replacing the static `ShellInfo.job` variable.

## Changes
### User-facing changes
- Adds of all PCT assets, components, and skill/resource/state code (nothing exposed right now since ShellJob is set to BLM here)
- Adds CN localization for PCT buff names (appears when mousing over a buff icon)
- Adds the DET stat in configuration
- Adds the `InCombat` resource
### Internal-only refactors
- Adds the `PaintGaugeCounter` and `ResourceText` job gauge components
- Adds the ability to display buffs that have no timer and last indefinitely
- Separates IntroSection text based on the `ShellJob` variable
- Changes `skillIconImages` map to be populated lazily--our existing approach made some assumptions about webpack import order that caused issues when trying to run the PCT sanity test I wrote
- **Adds a normalization pass to skill names:** I previously made some capitalization errors in skill names ("Thunder **I**n Red" rather than "Thunder in Red") that I had to fix to support export to Amarantine's sim, so I added a normalization pass to all skill name loads. This means **skill names in serialized contexts (fight records, sequence presets) are now case-insensitive**--"Fire 3" and "fiRE 3" are now the same. If we ever want to rename "Fire 4" to match the in-game "Fire IV", that can be handled through this normalization layer as well.
- `DamageStatistics::expandNode` now checks if a skill had computed potencies, instead of checking against an explicit whitelist of non-damaging skills.
- Adds a `useStackWithRecast` function to support abilities with variable base recast times

## Other notes
BLM in the Shell uses "/ffxiv-blm-rotation/" at the end of its URL while PCT in the Shell does not,
which matters when specifying the URL for preset markers. This is controlled by a variable for now.

@miyehn please double check the text in `IntroSection`, since I moved a lot of it into variables to allow different text in PCT/BLM (though a lot of it will end up becoming shared when the deployments merge). I think I copied the text from an up-to-date version of the site, but I might've missed some changes since git can't really capture the diffs effectively when variables are moved like this.

miyehn commented on src/Game/Resources.ts:
```diff
@@ -105,25 +105,37 @@ export class DoTBuff extends Resource {
 export class CoolDown extends Resource {
 	readonly #cdPerStack: number;
 	#recastTimeScale: number;
+	#currentBaseRecast: number;
```
Then I think we can get rid of the `#cdPerStack` above..?
Or `#cdPerStack` can be renamed to, say, `#defaultBaseRecast`, then restore `#currentBaseRecast` to that when some special recast is over.. somehow?

zqsz-xiv replied:
Yeah, the changes I made to support variable GCD are pretty janky. Right now this just relies on `useStackWithRecast` (rather than `useStack`) being called every time for this particular cooldown. Maybe the correct move is to just make a separate child class for the GCD to encode this behavior.

That said, I don't think refactoring this is super urgent, since it's buried beneath `GameState::useSpell` and `GameState::useAbility`, so other job writers shouldn't need to interact with this at all. Off the top of my head, I don't think there's any non-GCD abilities that have _variable_ cooldowns based on current state, just reductions of an existing cooldown (Tempera Coat popping for PCT, Mage's Ballad resetting the oGCD thing for BRD, unmend reducing ~~Plunge~~ the gapcloser CD on DRK).

zqsz-xiv replied:
Renamed to `#defaultBaseRecast` as suggested. Resetting the value of `#currentBaseRecast` after a `useStackWithRecast` call might break some things, and for the reasons in my last comment I don't think it's a change we need to make any time soon.

miyehn commented on src/Game/Resources.ts:
```diff
@@ -105,25 +105,37 @@ export class DoTBuff extends Resource {
+	useStackWithRecast(game: GameState, recast: number) {
+		// roll the GCD with a special recast value
+		this.consume(this.#currentBaseRecast);
+		this.maxValue = recast;
```
I guess we can also get rid of `useStack` above and make `recast` an optional arg..
But also setting `maxValue` means it's assuming this cooldown has 1 stack maximum. Which is true for now, seeing this function is only used for spells, but I'm still not sure we should make this assumption... :0

miyehn reviewed (commented):
[WIP]
I haven't looked at everything yet just taking notes as I go so there could be comments that don't make sense or questions that are answered by the rest of the changes.. don't mind me xD

--
Is this expected? Auto CDH skill potencies don't match :0 My first intuition is to at least put the multiplier from CDH there (for example 1.71 something from whatever config I'm looking at), if not showing a complete breakdown in case anyone wants to check the math behind.
But having a seemingly magical multiplier number probably won't be as useful as modifiers from AF/UI anyway. So I'm also not sure what's the best option
<img width="881" alt="image" src="https://github.com/user-attachments/assets/95f37871-e1ac-4cf2-b1e6-13a67cd87688">

zqsz-xiv commented on src/Game/XIVMath.ts:
```diff
@@ -77,6 +99,9 @@ export class XIVMath {
 		if (buff === ResourceType.LeyLines) {
 			return 15;
 		}
+		if (buff === ResourceType.Inspiration) {
```
Inspiration doesn't apply to motifs and Rainbow Drip, but does apply to all the fillers/subtractive + holy/comet. I leave the responsibility of setting the active buff up to the caller, in a function in `PCTState`:
https://github.com/xivintheshell/xivintheshell/blob/42e63af977fe663d484d9bb962dcd41d939e0875/src/Game/Jobs/PCT.ts#L83-L86

zqsz-xiv commented on src/Game/Common.ts:
```diff
@@ -19,6 +19,7 @@ export const enum Aspect {
 	Fire = "Fire",
 	Ice = "Ice",
 	Lightning = "Lightning",
+	Hammer = "Hammer",
```
> I think Aspect is pretty specific to BLM too and not sure other jobs should reuse this enum

Definitely agree, I just used this as a quick hack when I was throwing together the PCT fork. I'd like to refactor this at some point to just let skills add special buffs/modifiers more easily.

> reading from the global ShellInfo.job (in getPotencyModifersFromResourceState and elsewhere) means the behavior of this function is going to be the same for every timeline slot

We'll need some further refactors to de-global this, but in the worst case we can just use `controller.game.job` instead of `ShellInfo.job`.

miyehn reviewed (commented):
[more WIP]

zqsz-xiv commented on src/Components/DamageStatistics.tsx:
```diff
@@ -149,6 +154,12 @@ function BuffTag(props: {buff?: PotencyModifierType, tc?: boolean}) {
 		color = colors.resources.enochian;
+	} else if (props.buff === PotencyModifierType.HAMMER) {
+		text = localize({en: "CDH"});
+		color = colors.resources.astralFire;
```
Done! We should probably split up the class colors into separate objects at some point.

zqsz-xiv commented on src/Controller/DamageStatistics.ts:
```diff
@@ -171,36 +155,50 @@ function expandNode(node: ActionNode) : ExpandedNode {
-					break;
+					res.displayedModifiers.push(tag);
+					res.calculationModifiers.push(modifier);
+					// TODO check why blm does a break; here
```
Fixed, and assigned `mainPotency.modifiers` outside the loop.

I'm a bit confused as to why just doing `res.calculationModifiers = mainPotency.modifiers` still excludes pot from the calculation, but the displayed potencies match those in current main.

miyehn replied:
iirc pot just has a modifier of 1 so it's not shown in calculations. And that is because pot is not applied to every instance of the skills in a row, and it wouldn't make much sense to make a new row just to separate the same skill with and without pot.
If you think it'd be helpful we can always add another line in the calculation tooltip and hypothetically include pot as well

miyehn replied:
lmao I think I just also answered my own question of why CDH and other party buffs can't/shouldn't be part of the calculation here 😅

zqsz-xiv replied:
I think auto-crit/auto-CDH/auto-DH skills would make sense in general to include as a separate tooltip; off the top of my head I know MCH's Reassemble will make the next weaponskill auto-CDH, so it would make sense to display that distinctly. I'm too lazy to figure out how to make it show up right now though.

zqsz-xiv commented:
Addressed comments, and also added the `InCombat` resource override (necessary for PCT and possibly other jobs). After this is merged I'm gonna add a couple small refactors:
- add a second check of the MP requirement when a cast hits its snapshot window (this fixes the eno drop bug I mentioned in discord)
- split color configuration object up by job
- move `Game/StatsModifier.ts` logic into `Game/Jobs/BLM.ts`
- split logic for highlighting abilities on hotbar to individual skills, rather than a giant chain in `GameState::getSkillAvailabilityStatus`

zqsz-xiv commented:
@miyehn did you want to look at anything else before merging?

miyehn commented:
I rose from the dead... Today I will be catching up and finishing things I left off. Apologies for the delay \_(:з」∠)_

miyehn commented on src/Components/IntroSection.tsx:
```diff
@@ -27,6 +42,82 @@ function Changelog() {
+						<li><b>Eshiya (Galahad Donnadieu @ Exodus)</b>: the PM and the big brain BLM</li>
+						<li><b>miyehn (Ellyn Waterford @ Sargatanas)</b>: software developer and a humble BLM student</li>
+						<li><b>Turtle</b>, who did the heavy work of updating this tool from Endwalker to Dawntrail</li>
+						<li><b>Yara, Spider, Santa, shanzhe, Akairyu</b>, and many other players who contributed feature suggestions, timeline markers, bug reports, etc.</li>
```
Would you want to / do you mind taking up another line above this (people listed in the order they came and made significant contribution to the development of this tool)? You can make the change or I can do it in my next commit

zqsz-xiv replied:
Sure! I added a line for myself (no zh/jp localization), feel free to localize yourself.

miyehn reviewed (approved):
Alrighty looks all good to me. Tysm let's gooo 🚀

zqsz-xiv commented on src/Game/GameState.ts:
```diff
@@ -337,6 +386,15 @@ export abstract class GameState {
 				}
 			));
+
+			if (potency) {
```
It's the damage application time. I coded it that way in the initial fork because it seemed to more or less match the in-game behavior, and nobody has raised an issue with it yet.
